File: swift-bridge/Sources/SmixSDK/Smix.swift
// Smix top-level entry — wires `Smix.launchApp` over the FFI driving
// surface. Constructs an `SmixDriver` pointed at the local runner, opens
// a session for the target bundle id, launches it, and hands back an
// `App` holding both handles.

import Foundation
import SmixCoreFFIBindings
import SmixRunnerCore

/// Top-level entry to the smix SDK. Launches and attaches to iOS
/// Simulator apps over the FFI boundary to a running smix runner.
public enum Smix {
    /// Default local runner port (mirror `RunnerPortResolver.defaultPort`).
    public static let defaultRunnerPort: UInt16 = RunnerPortResolver.defaultPort

    /// Launch the target app on the simulator and return an [`App`]
    /// handle.
    ///
    /// Constructs an [`SmixDriver`] on `port`, opens a session bound to
    /// the target bundle id, launches the app, and returns an [`App`]
    /// holding the driver (tree / list) and session (act).
    ///
    /// - Parameters:
    ///   - target: bundle identifier to launch (must already be
    ///     installed on the sim).
    ///   - port: local runner port. Defaults to
    ///     [`Smix.defaultRunnerPort`].
    /// - Returns: an [`App`] handle bound to the launched process.
    public static func launchApp(
        _ target: AppTarget,
        port: UInt16 = Smix.defaultRunnerPort
    ) async throws -> App {
        switch target {
        case let .bundleId(id):
            let driver = SmixDriver(port: port)
            let session = try await driver.openSession(bundleId: id, cancel: nil)
            try await session.launchApp(cancel: nil)
            return App(bundleId: id, driver: driver, session: session)
        }
    }
}

/// What to launch — a registered bundle identifier already installed on
/// the sim.
public enum AppTarget: Sendable, Equatable {
    /// e.g. `com.example.MyApp` — must already be installed on sim.
    case bundleId(String)
}

/// SDK error category used for compile-time-known unimplemented
/// surface, distinct from [`ExpectationFailure`] (runtime resolution
/// failure).
public enum SmixError: Error, Sendable, Equatable {
    /// Method exists on the API surface but body is not yet wired.
    case notImplemented(stage: String, api: String)
}

File: android-runner/sdk/src/main/kotlin/dev/smix/sdk/Smix.kt
// SmixSDK Kotlin top-level entry.
//
// Mirrors swift-bridge/Sources/SmixSDK/Smix.swift. Playwright-style
// ergonomic facade backed by UniFFI Kotlin bindings (uniffi.smix.*).
//
// Usage:
//   val app = Smix.launchApp(AppTarget.BundleId("com.example.MyApp"))
//   app.tap(Selector.Id("btn-login"))
//   app.find(Selector.Text("Welcome")).toBeVisible(timeout = 5.seconds)
//
// This module is intended for test targets only
// (androidTestImplementation / debugImplementation); there is no linkage
// path into an app's release build.

package dev.smix.sdk

/**
 * Top-level entry to the smix SDK. Launches and attaches to Android
 * emulator apps over the FFI boundary to a running smix runner.
 */
object Smix {
    /**
     * Default local runner port. The Android runner listens on 28080 by
     * convention (see docs/ai-guide/05-cli.md; iOS uses 22087).
     */
    val defaultRunnerPort: UShort = 28080u

    /**
     * Launch the target app on the emulator and return an [App] handle.
     *
     * Constructs a [Driver] on [port], opens a session bound to the
     * target bundle id, launches the app, and returns an [App] holding
     * the driver (tree / list) and session (act).
     */
    suspend fun launchApp(
        target: AppTarget,
        port: UShort = defaultRunnerPort,
        resolver: SelectorResolver = DefaultFfiResolver,
        labelsResolver: LabelResolver = DefaultFfiLabelsResolver,
    ): App = when (target) {
        is AppTarget.BundleId -> {
            val driver = FfiDriver(port)
            val session = driver.openSession(target.value)
            session.launchApp()
            App(
                bundleId = target.value,
                driver = driver,
                session = session,
                resolver = resolver,
                labelsResolver = labelsResolver,
            )
        }
    }
}

/**
 * What to launch — a registered bundle identifier already installed on
 * the emulator.
 */
sealed interface AppTarget {
    data class BundleId(val value: String) : AppTarget
}

/**
 * SDK error category for compile-time-known unimplemented surface,
 * distinct from [ExpectationFailure] (runtime resolution failure).
 */
sealed class SmixError(message: String) : Exception(message) {
    class NotImplemented(val stage: String, val api: String) :
        SmixError("$api not implemented yet ($stage)")
}
